third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -3236,6 +3243,156 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
+            if name_str == "vector" || name.addr().is_one() {
+                let module_env = self.parent.parent.env.get_module(*mid);
+                for fenv in module_env.get_functions() {
+                    if fenv.get_full_name_str() == target_str {
+                        let symbol = fenv.get_name();
+                        symbol_opt = Some(symbol);
+                        module_opt = Some(name.clone());
+                        break;
+                    }
+                }
+                break;
+            }
+        }
+        if let (Some(symbol), Some(module_name)) = (symbol_opt, module_opt) {
+            let name_loc = &self.to_loc(&e.loc);
+            return self.translate_call(
+                loc,
+                name_loc,
+                CallKind::Regular,
+                &Some(module_name.clone()),
+                symbol,
+                &None,
+                &[
+                    &sp(e.loc, EA::Exp_::Borrow(mutable, Box::new(e.clone()))),
+                    i,
+                ],
+                expected_type,
+                context,
+            );
+        }
+        ExpData::Invalid(self.env().new_node_id())
+    }
+
+    fn is_struct_or_schema(&self, maccess: &EA::ModuleAccess) -> (bool, bool) {
+        let global_var_sym = match &maccess.value {
+            EA::ModuleAccess_::ModuleAccess(..) => self.parent.module_access_to_qualified(maccess),
+            EA::ModuleAccess_::Name(name) => {
+                let sym = self.symbol_pool().make(name.value.as_str());
+                self.parent.qualified_by_module(sym)
+            },
+        };
+        (
+            self.parent
+                .parent
+                .struct_table
+                .contains_key(&global_var_sym),
+            self.parent
+                .parent
+                .spec_schema_table
+                .contains_key(&global_var_sym),
+        )
+    }
+
+    /// Try to translate a resource or vector Index expression
+    fn try_resource_or_vector_index(
+        &mut self,
+        loc: &Loc,
+        target: &EA::Exp,
+        index: &EA::Exp,
+        context: &ErrorMessageContext,
+        ty: &Type,
+        mutable: bool,
+        check_language_version: bool,
+    ) -> Option<ExpData> {
+        let mut call = None;
+        if let EA::Exp_::Name(m, _) = &target.value {
+            let (is_struct, is_schema) = self.is_struct_or_schema(m);
+            if is_struct {
+                self.check_language_version(loc, "resource indexing", LanguageVersion::V2_0);
+                call = Some(self.call_to_borrow_global(loc, target, index, mutable, ty, context));
+            } else if is_schema && self.env().language_version == LanguageVersion::V2_0 {
+                self.error(loc, "indexing is only allowed for resource or vector");
+                call = Some(self.new_error_exp());
+            }
+        }
+        if !self.is_spec_mode() {
```

Comment:
Should this be an `else-if` for the `if-let`, for clarity (to make sure that we could not have entered `is_struct` as well as this branch)?